boa_inputgen/src/ident_walk.rs

```diff
@@ -0,0 +1,517 @@
+//! Identifier and symbol walker for ensuring that generated inputs do not fail with "string
+//! disappeared". The AST is walked by level-order traversal and Sym instances are replaced with
+//! valid ones generated by arbitrary.
+
+use boa_engine::syntax::ast::{
+    node::{
+        declaration::{BindingPatternTypeArray, BindingPatternTypeObject},
+        iteration::IterableLoopInitializer,
+        object::{MethodDefinition, PropertyDefinition, PropertyName},
+        operator::assign::AssignTarget,
+        template::TemplateElement,
+        AsyncFunctionExpr, AsyncGeneratorExpr, Block, Declaration, DeclarationList,
+        DeclarationPattern, FormalParameter, FunctionExpr, GeneratorExpr, Identifier,
+        StatementList,
+    },
+    Const, Node,
+};
+use boa_interner::Sym;
+
+/// Given this list of syms, walk the AST provided and replace any syms with a matching sym from
+/// the list.
+pub(crate) fn replace_syms(syms: &[Sym], sample: &mut StatementList) {
+    replace_inner(syms, sample.items_mut().iter_mut().map(extendo).collect());
+}
+
+/// Extend the lifetime of an arbitrary reference temporarily. Only safe in this context because
+/// we only hold these references during the walk, so they are *relatively* static. Holding
+/// references to the AST while continuing to walk makes a double-mutable-reference scenario which
+/// the borrow checker denies, but it's entirely safe since we only modify one Sym at a time and the
+/// modification is idempotent. We don't need to pin because no data structures are modified, only
+/// particular values.
+fn extendo<T>(node: &mut T) -> &'static mut T {
+    unsafe { &mut *(node as *mut T) }
+}
```

Comment:
Maybe it makes sense to make the function `unsafe`, and then use the `unsafe` block where it's used, to make sure that each call is safe (so, invariants are checked on each call)